tc/core/compiler-inl.h

```diff
@@ -0,0 +1,101 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "tc/core/compiler.h"
+
+#include <sstream>
+#include <string>
+
+#include "tc/core/flags.h"
+#include "tc/core/halide_utils.h"
+#include "tc/core/tensor.h"
+#include "tc/lang/canonicalize.h"
+
+namespace tc {
+namespace {
+void checkInputsCompliant(
+    const tc2halide::HalideComponents& halideComponents,
+    const std::vector<const DLConstTensor*>& inputsInfo) {
+  if (inputsInfo.size() != halideComponents.inputs.size()) {
+    throw lang::ErrorReport(halideComponents.getDef())
+        << "expected " << halideComponents.inputs.size() << " inputs but found "
+        << inputsInfo.size();
+  }
+  for (size_t i = 0; i < inputsInfo.size(); ++i) {
+    auto dltype = inputsInfo[i]->dtype;
+    auto htype = halideComponents.inputs[i].type();
+    // we have three type representations here: (1) halide Type (2) DLTensor
+    // type, and (3) the token representing the type in the frontend (e.g.
+    // TK_FLOAT) we need to translate to (3) to report user facing errors
+    auto dltype =
+        lang::TypeInfo(lang::TypeInfo::Code(dltype.code), dltype.bits)
+            .toScalarToken();
+    auto htype =
+        lang::TypeInfo(lang::TypeInfo::Code(htype.code()), htype.bits())
+            .toScalarToken();
+    if (dltype != htype) {
+      throw lang::ErrorReport(halideComponents.getDef().params()[i])
+          << "expected type " << lang::kindToString(htype) << " but found "
+          << lang::kindToString(dltype);
+    }
+    int hdim = halideComponents.inputs[i].dimensions();
+    int dldim = inputsInfo[i]->ndim;
+    if (dldim != hdim) {
+      throw lang::ErrorReport(halideComponents.getDef().params()[i])
+          << "expected a tensor with " << hdim << " dimensions but found "
+          << dldim << " dimensions.";
+    }
+  }
+}
+} // namespace
+template <typename Backend>
+std::unique_ptr<typename Backend::ExecutorType> compile(
+    const std::string& tc,
+    const std::string& entryPoint,
+    const std::vector<const DLConstTensor*>& inputs,
+    /* TODO: in the future also pass outputs for stride and alignment info */
+    const typename Backend::MappingOptionsType& options) {
+  auto parsedTcs = detail::parse(tc);
+  CHECK_EQ(1, parsedTcs.count(entryPoint))
+      << "attempting to access undefined function " << entryPoint;
+  return compile<Backend>(parsedTcs[entryPoint], inputs, options);
+}
+
+template <typename Backend>
+std::unique_ptr<typename Backend::ExecutorType> compile(
+    lang::TreeRef tcDefinition,
+    const std::vector<const DLConstTensor*>& inputs,
+    /* TODO: in the future also pass outputs for stride and alignment info */
+    const typename Backend::MappingOptionsType& options) {
+  using CompilationResultType = typename Backend::CompilationResultType;
+
+  auto inputsInfo = makeTensorInfoVector(inputs);
+  auto outputsInfo = detail::inferOutputTensorInfo(tcDefinition, inputs);
```

Comment:
so we are going to infer it two times. Once here and once in places like TcOp. Do you think we can have an interface which allows to get this from the get go and avoid output shape inference twice?